Add API logging for interpolate()

NicolasHug opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88212


This PR calls `torch._C._log_api_usage_once()` for `torch.nn.functional.interpolate()`. See [this internal post](https://fb.workplace.com/groups/1144215345733672/permalink/2382133531941841/) for more context about the level of details needed in the logger


cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

NicolasHug committed:
Add API logging for interpolate()

[ghstack-poisoned]

albanD commented on torch/nn/functional.py:
```diff
@@ -3916,6 +3916,12 @@ def interpolate(input: Tensor, size: Optional[int] = None, scale_factor: Optiona
+    # Logging logic
+    is_channels_last = input.is_contiguous(memory_format=torch.channels_last)
+    is_image_or_mask = input.ndim == 4 and input.shape[1] < 4
+    log_string = f"torch.nn.functional.interpolate_dtype={input.dtype}_mode={mode}_antialias={antialias}_channelslast={is_channels_last}_imageormask={is_image_or_mask}"
```
I don't think this is ok. Creating such string is quite expensive and I don't think we want all our users to pay that price.

NicolasHug replied:
I understand this may not be desirable on more ubiquitous ops like `sum()` or `max()`, but is this a problem in practice for `interpolate()` which is in general a fairly slow operation? On a typical input the string construction would only account for 0.1% of the `interpolate()` call:

```py
[ins] In [13]: import torch

[ins] In [14]: input = torch.rand(1, 3, 350, 280); antialias = True; mode = "bilinear"

[ins] In [16]: torch.set_num_threads(1)

[ins] In [17]: %timeit torch.nn.functional.interpolate(input, size=(224, 224), mode=mode, antialias=antialias)
1.05 ms ± 449 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

[ins] In [18]: %%timeit
          ...:     is_channels_last = input.is_contiguous(memory_format=torch.channels_last)
          ...:     is_image_or_mask = input.ndim == 4 and input.shape[1] < 4
          ...:     log_string = f"torch.nn.functional.interpolate_dtype={input.dtype}_mode={mode}_antialias={antialias}
          ...: _channelslast={is_channels_last}_imageormask={is_image_or_mask}"
          ...: 
1.15 µs ± 1.97 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

If you feel strongly against this, would you mind sharing your thoughts on [this thread](https://fb.workplace.com/groups/1144215345733672/posts/2382133531941841/?comment_id=2382227655265762&reply_comment_id=2382864245202103)? According to our internal discussion with @jisaacso  and @colin2328 , this kind of logging should be possible?

albanD replied:
Sent there.

albanD reviewed (approved):
From internal discussion, this is an ok short term solution while we are building a better infra for this.
This will be removed when the new infra is in place.

NicolasHug committed:
Update on "Add API logging for interpolate()"



This PR calls `torch._C._log_api_usage_once()` for `torch.nn.functional.interpolate()`. See [this internal post](https://fb.workplace.com/groups/1144215345733672/permalink/2382133531941841/) for more context about the level of details needed in the logger


cc albanD mruberry jbschlosser walterddr kshitij12345 saketh-are

[ghstack-poisoned]

dzhulgakov commented on torch/nn/functional.py:
```diff
@@ -3916,6 +3916,15 @@ def interpolate(input: Tensor, size: Optional[int] = None, scale_factor: Optiona
+            f"torch.nn.functional.interpolate_dtype={input.dtype}_mode={mode}_antialias={antialias}_"
+            f"channelslast={is_channels_last}_imageormask={is_image_or_mask}"
+    )
+    torch._C._log_api_usage_once(log_string)
```
It's doing a uniqueness check for the log message, so definitely has some overhead: https://github.com/pytorch/pytorch/blob/master/torch/csrc/Module.cpp#L1159

What is the need for this particular logging? If you're trying to capture usage at scale, you could probably look at ATen level operator logging (which at Meta is hooked up to Scuba)

NicolasHug replied:
Thanks for chiming in @dzhulgakov . The reason for this logging is to identify which low-level kernels are more often used internally (and by whom), so we can prioritize perf improvements. Would you mind sharing more context about the Aten operator logging you're mentioning? Are you saying that all the calls to `torch._C._nn.XYZ` like [these ones](https://github.com/pytorch/pytorch/blob/dc00bb51b8d370bf3891f0edb2c6e0c2914e329a/torch/nn/functional.py#L3920-L3958)  are logged somewhere internally?

dzhulgakov replied:
Yes, all aten operators are getting logged with low sampling probability. It's done by installing an observer like described here: https://pytorch.org/docs/stable/notes/large_scale_deployments.html#fleet-wide-operator-profiling . At Meta look up `pytorch_operator_stats` in Scuba.

NicolasHug committed:
Update on "Add API logging for interpolate()"



This PR calls `torch._C._log_api_usage_once()` for `torch.nn.functional.interpolate()`. See [this internal post](https://fb.workplace.com/groups/1144215345733672/permalink/2382133531941841/) for more context about the level of details needed in the logger


cc albanD mruberry jbschlosser walterddr kshitij12345 saketh-are

[ghstack-poisoned]
